openwisp_controller/config/admin.py

```diff
@@ -849,6 +867,25 @@ class Media(BaseConfigAdmin):
 
 
 class DeviceGroupForm(BaseForm):
+    _templates = None
+
+    def clean_templates(self):
+        templates = self.cleaned_data.get('templates')
+        self._templates = templates
+        return templates
+
+    def save(self, *args, **kwargs):
+        instance = super().save(*args, **kwargs)
+        if not self.instance._state.adding:
+            DeviceGroup.templates_changed(
```

Comment:
During my testing, I see that this code gets called also when the templates are not changed at all, which is bad, because on a system with thousands of devices we will generate many queries and after each change a network request for the configuration update will also be generated, causing unnecessary load.

Why aren't you sending this signal only if templates have been changed?